File: src/main/java/shopping/wish/service/WishService.java
package shopping.wish.service;

import org.springframework.stereotype.Service;
import shopping.common.exception.ProductNotFoundException;
import shopping.common.exception.WishListNotFoundException;
import shopping.product.domain.Product;
import shopping.product.repository.ProductRepository;
import shopping.wish.domain.WishList;
import shopping.wish.dto.WishProductResponse;
import shopping.wish.repository.WishListItemRepository;
import shopping.wish.repository.WishListRepository;

import java.util.List;

@Service
public class WishService {

    private final WishListRepository wishListRepository;
    private final WishListItemRepository wishListItemRepository;
    private final ProductRepository productRepository;

    public WishService(
            WishListRepository wishListRepository,
            WishListItemRepository wishListItemRepository,
            ProductRepository productRepository
    ) {
        this.wishListRepository = wishListRepository;
        this.wishListItemRepository = wishListItemRepository;
        this.productRepository = productRepository;
    }

    public void addWish(Long memberId, Long productId) {

        WishList wishList = wishListRepository.findByMemberId(memberId)
                .orElseThrow(WishListNotFoundException::new);

        Product product = productRepository.findById(productId)
                .orElseThrow(ProductNotFoundException::new);

        wishList.addItem(product);

        wishListRepository.save(wishList);
    }

    public void deleteWish(Long memberId, Long wishId) {

        wishListItemRepository.deleteById(wishId);
    }

    public List<WishProductResponse> getWishList(Long memberId) {

        WishList wishList = wishListRepository.findByMemberId(memberId)
                .orElseThrow(WishListNotFoundException::new);

        return wishListItemRepository.findByWishListId(wishList.getId())
                .stream()
                .map(WishProductResponse::from)
                .toList();
    }
}

File: src/main/kotlin/shopping/application/wish/WishService.kt
package shopping.application.wish

import shopping.domain.product.ProductRepository
import shopping.domain.wish.Wish
import shopping.domain.wish.WishRepository
import shopping.support.error.CoreException
import shopping.support.error.ErrorType

class WishService(
    private val wishRepository: WishRepository,
    private val productRepository: ProductRepository
) {
    fun addWish(memberId: Long, productId: Long): Wish {
        if (!productRepository.existsById(productId)) {
            throw CoreException(ErrorType.NOT_FOUND, "상품이 존재하지 않습니다.")
        }
        if (wishRepository.existsByMemberIdAndProductId(memberId, productId)) {
            throw CoreException(ErrorType.CONFLICT, "이미 위시리스트에 추가된 상품입니다.")
        }
        return wishRepository.save(Wish(memberId = memberId, productId = productId))
    }

    fun removeWish(memberId: Long, wishId: Long) {
        val wish = wishRepository.findByMemberId(memberId)
            .find { it.id == wishId }
            ?: throw CoreException(ErrorType.NOT_FOUND, "위시리스트에 존재하지 않는 항목입니다.")
        wishRepository.deleteById(wish.id)
    }

    fun getWishes(memberId: Long): List<Wish> {
        return wishRepository.findByMemberId(memberId)
    }
}
